### ISSUE TITLE
Failure on test_wbemcli.py on RHEL 7.3

### ISSUE THREAD

**Issue by andy-maier** [pywbem/pywbem#849]
I get this test error for test_wbemcli.py on my RHEL 7.3 system. It occurs on every testcase in the same way, and is probably all the same reason. It succeeds on my Ubuntu 14.04 system at home.

```
=================================== FAILURES ===================================
______________________ TestsContainer.test_def_namespace _______________________

self = <test_wbemcli.TestsContainer testMethod=test_def_namespace>

    def test(self):  # pylint: disable=missing-docstring
        """ The test method that is generated."""
        # create the path for the quit script. Required to
        # exit wbemcli after the test.
        quit_script_file = create_abs_path('wbemcli_quit_script.py')
    
        script_name = 'wbemcli.bat' if os.name == 'nt' else 'wbemcli'
    
        url = 'http://blah' if test_params.url is None \
            else test_params.url
    
        cmd = ('%s %s %s -s %s' % (script_name,
                                   url,
                                   test_params.cmd,
                                   quit_script_file))
    
        # Disable python warnings for wbemcli call
        # because some imports generate deprecated warnings
        # that appear in std_err when nothing expected
        bash_cmd = 'bash -c "set -o pipefail;  PYTHONPATH=. ' \
                   'PYTHONWARNINGS= %s"' % cmd
    
        proc = Popen(bash_cmd, shell=True, stdout=PIPE, stderr=PIPE)
        std_out, std_err = proc.communicate()
        exitcode = proc.returncode
    
        # print('bash_cmd: %s\nexitcode: %s std_out:\n%s\nstd_err:'
        #      '\n%s' % (bash_cmd, exitcode, std_out, std_err))
    
        if six.PY3:
            std_out = std_out.decode()
            std_err = std_err.decode()
    
        if test_params.expected_exitcode is not None:
            self.assertEqual(exitcode, test_params.expected_exitcode,
                             ('Test %s: Unexpected ExitCode Err. '
                              'Expected %s: cmd="%s": '
                              'exitcode %s: stderr=%s' %
                              (test_name, test_params.expected_exitcode,
                               cmd,
>                              exitcode, std_err)))
E           AssertionError: Test test_def_namespace: Unexpected ExitCode Err. Expected 0: cmd="wbemcli http://blah  -s /home/maiera/Projects/pywbem/repos/pywbem/testsuite/wbemcli_quit_script.py": exitcode 126: stderr=bash: /home/maiera/virtualenvs/pywbem27/bin/wbemcli: python: bad interpreter: No such file or directory

testsuite/test_wbemcli.py:167: AssertionError
```

**Comment by andy-maier** [comment]
I was digging a bit into this, and to be on the safe side, I reinstalled the whole virtualenv.
The pywbem package from the working directory is installed as a site package in the virtualenv, and that was done using "make install" which performs a "python setup.py install".

That installation of pywbem creates a "wbemcli" script in the bin directory of the virtualenv:
```
$ cat ~/virtualenvs/pywbem27c/bin/wbemcli
#!/home/maiera/virtualenvs/pywbem27c/bin/python
# EASY-INSTALL-SCRIPT: 'pywbem==0.12.0.dev0','wbemcli'
__requires__ = 'pywbem==0.12.0.dev0'
__import__('pkg_resources').run_script('pywbem==0.12.0.dev0', 'wbemcli')
```
which calls the "wbemcli" script that we have in the main directory after some lookup process. That lookup process raises an exception, if that script is invoked with `PYTHONPATH=.` (and only then):
```
(pywbem27c)[...]$ which wbemcli
~/virtualenvs/pywbem27c/bin/wbemcli

(pywbem27c)[...]$ wbemcli
usage: wbemcli [options] server
wbemcli: error: No WBEM server specified

(pywbem27c)[...]$ PYTHONPATH=. wbemcli
Traceback (most recent call last):
  File "/home/maiera/virtualenvs/pywbem27c/bin/wbemcli", line 4, in <module>
    __import__('pkg_resources').run_script('pywbem==0.12.0.dev0', 'wbemcli')
  File "/home/maiera/virtualenvs/pywbem27c/lib/python2.7/site-packages/pkg_resources/__init__.py", line 750, in run_script
    self.require(requires)[0].run_script(script_name, ns)
  File "/home/maiera/virtualenvs/pywbem27c/lib/python2.7/site-packages/pkg_resources/__init__.py", line 1516, in run_script
    raise ResolutionError("No script named %r" % script_name)
pkg_resources.ResolutionError: No script named 'wbemcli'
```

This error happens during "make test" in my original post.

I was googling for occurrences of this error, but I did not find any detailed information as to why it fails or what can fix it, other than "reinstall the package". Which I did.

`pkg_resources` is not a pip/pypi package on its own, but a Python module/package that comes with (or at least is updated by) the `setuptools`  pip/pypi package.

**Comment by andy-maier** [comment]
I tried to circumvent the issue by running `python setup.py egg_info` which creates a local `pywbem.egg-info` metadata directory. I verified by debugging that pkg_resources has that directory in `self.egg_info` on line 1490 of the code shown above when checking for the script file (when using `PYTHONPATH=.`).

However, that local egg metadata directory does not contain the `scripts` subdirectory, which is why it fails.

So I tried to understand what `python setup.py install` does,because it obviously creates the egg metadata directory **with** the scripts in the virtualenv. It turns out that it also creates the local egg metadata directory via the `egg_info` command, but another command creates the scripts somewhere under `build` and creates a .egg file (of zip format) in the `dist` directory. The installation into the virtualenv is using that .egg file. So locally, there is never a egg metadata directory that contains the scripts.

Bottom line, I don't know how setup.py could be used to produce a local egg metadata directory that has the scripts inside.

I still don't understand why this (i.e. `PYTHONPATH=. wbemcli`) ever worked, and why it still works on Ubuntu. We should debug this on Ubuntu to see what egg metadata directory it has when making the check on line 1490, above.

**Comment by andy-maier** [comment]
I found that when installing pywbem with `pip install . -e`, issuing `PYTHONPATH=. wbemcli` works.

That might be the way to go, at least as a circumvention. We would simply change the way pywbem is installed in the makefile, and make test would no longer need to use `PYTHONPATH=.` with `py.test`, because the latest files from any modules in the working directory are used on behalf of the `-e` option of `pip`. The test files are searched in the current subtree anyway, and therefore do not depend on `PYTHONPATH=.` anyway.

**Comment by andy-maier** [comment]
I created PR #851 with the circumvention discussed in the previous comment.

**Comment by DreadPirateShawn** [comment]
This write-up is super helpful for explaining that the current error
```
pkg_resources.ResolutionError: Script 'scripts/foo' not found in metadata at '/my-project/my-project.egg-info'
```
can be triggered by `PYTHONPATH=/my-project` causing the script to be searched for in the wrong location.

(Adding the new error message for search reasons, as currently the error message is "clearer" but still incredibly confusing to suss out, without this ticket's context.)
